### PR TITLE
Persist the code version when fedimintd starts up

### PR DESCRIPTION
**Pull request by m1sterc001guy**
In preparation for [Database Migrations](https://github.com/fedimint/fedimint/pull/1229/files), we will need a mechanism for determining when the code version of fedimintd has changed. We already have the code version in the config file, this PR modifies fedimintd so that value is persisted in the database.

Currently, if the code version already exists in the database, we don't do anything and simply output a log message. This is when we'll need to perform the migration. If the code version is not in the database, we persist it. If it is in the database and it matches the current code version, we also just log and continue.

Used dbdump and tmuxinator to verify that the value is persisted correctly.

### CODE REVIEW

File: fedimint-dbdump/src/main.rs [hunk 1]
```diff
@@ -612,6 +616,29 @@ impl<'a> DatabaseDump<'a> {
         self.serialized
             .insert("Client".to_string(), Box::new(client));
     }
+
+    async fn get_database_data(&mut self) {
```
Comment (dpc) [inline fedimint-dbdump/src/main.rs:620]: It says `_get` but doesn't return anything, has no comment, I'm confused. :D

File: fedimintd/src/bin/main.rs [hunk 2]
```diff
@@ -37,6 +40,32 @@ pub struct ServerOpts {
     pub with_telemetry: bool,
 }
 
+async fn check_code_version(
```
Comment (dpc) [inline fedimintd/src/bin/main.rs:43]: Comment would be great. :P

File: fedimint-dbdump/src/main.rs (hunk 1 above)
Comment (m1sterc001guy) [reply fedimint-dbdump/src/main.rs:620]: Perhaps "retrieve" is a better name. It just reads it from the database and adds its to its internal map which is essentially an in-memory representation of the database. dump_database then serializes this. I can rename and add some comments.

**Comment by dpc** [comment]
BTW. I don't using git hashes as code versions is going to work well. We probably need to use bunch of counters (for each part of the code - modules will need their own counters), and update them each time a migration is required.

**Comment by justinmoon** [comment]
> BTW. I don't using git hashes as code versions is going to work well. We probably need to use bunch of counters (for each part of the code - modules will need their own counters), and update them each time a migration is required.

**Comment by justinmoon** [comment]
oops closing was accidental click

File: fedimintd/src/bin/main.rs [hunk 3]
```diff
@@ -37,6 +40,36 @@ pub struct ServerOpts {
...
+    mut dbtx: DatabaseTransaction<'_>,
+) -> Result<(), anyhow::Error> {
+    let persisted_code_version = dbtx.get_value(&DatabaseVersionKey).await?;
+
+    if let Some(persisted_code_version) = persisted_code_version {
+        if persisted_code_version == current_code_version {
+            tracing::info!("Code version matches database {}", persisted_code_version);
+        } else {
+            tracing::info!(
+                "Curernt code version {} does not match the database {}",
```
Comment (jkitman) [inline fedimintd/src/bin/main.rs:58]: Typo "Curernt"

**Review by jkitman — CHANGES_REQUESTED** [review]
I'm not sure I understand the plan for upgrades, although I wrote about it a little in #948.

Why do we need to persist the code version in the database when it is already persisted in the configuration file?

The update process could pause consensus, update the configs, migrate the database, then start the new server binary.

**Comment by m1sterc001guy** [comment]
> Why do we need to persist the code version in the database when it is already persisted in the configuration file?
> 
> The update process could pause consensus, update the configs, migrate the database, then start the new server binary.

How do we know which version we are migrating from if it has already been updated in the configuration? When upgrading the database values, we essentially need to explicitly define the transformations of key/values from old version to new versions. Today the code version is written during DKG. Are we planning on re-running DKG anytime there is an upgrade? If there's no consensus changes, that is probably too much overhead right?

**Comment by elsirion** [comment]
We should get on a call to come up with a unified vision for config and DB migration.

Having a sanity check won't hurt to not accidentally try to start the federation with a mismatching cfg/DB pair. Which one is the source of truth can be discussed, I'm slightly in favor of the config, in which case the DB could commit to the consensus config hash to check compatibility.

**Comment by jkitman** [comment]
@m1sterc001guy I see, I was envisioning an upgrade script that would handle making the modifications to the config and DB at the same time, guess it depends on how we decide to handle migrations.

**Review by jkitman — APPROVED** [review]

**Comment by m1sterc001guy** [comment]
From the call today: I believe we agreed that each module will need three version (consensus version, DB version, and API version). API version might be made a sub-version of consensus to ensure that clients can interact with them properly.

This PR currently takes the version from the config and writes it to the database. I think this is the wrong version, since that code version will be used to validate the consensus configs. I'm going to close this PR in favor of adding a database version to every module.

**Pull request closed without merging**